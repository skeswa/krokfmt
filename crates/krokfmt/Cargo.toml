[package]
name = "krokfmt"
version = "0.1.0"
edition.workspace = true
rust-version.workspace = true
authors.workspace = true
description = "A highly opinionated TypeScript code formatter"
license.workspace = true
repository.workspace = true
readme = "../../README.md"
keywords = ["typescript", "formatter", "code-formatter", "tsx", "jsx"]
categories = ["development-tools", "command-line-utilities"]

[[bin]]
name = "krokfmt"
path = "src/main.rs"

[dependencies]
# NOTE: we have to resort to some tomfoolery to get biome deps working. This is
# due mostly to the unwillingness of the maintainers to ensure their crates are
# stable.
#
# I yanked these version pins from https://github.com/brioche-dev/brioche/pull/184.
biome_console = "=0.5.7"
biome_deserialize = "=0.5.7"
biome_deserialize_macros = "=0.5.7"
biome_diagnostics = "=0.5.7"
biome_diagnostics_categories = "=0.5.7"
biome_diagnostics_macros = "=0.5.7"
biome_formatter = "=0.5.7"
biome_js_factory = "=0.5.7"
biome_js_formatter = "=0.5.7"
biome_js_parser = "=0.5.7"
biome_js_syntax = "=0.5.7"
biome_json_factory = "=0.5.7"
biome_json_parser = "=0.5.7"
biome_json_syntax = "=0.5.7"
biome_markup = "=0.5.7"
biome_parser = "=0.5.7"
biome_rowan = "=0.5.7"
biome_text_edit = "=0.5.7"
biome_text_size = "=0.5.7"
biome_unicode_table = "=0.5.7"
clap = { version = "4.5", features = ["derive"] }
swc_atoms = "^7.0.0"
swc_common = { version = "^14.0.0", features = ["tty-emitter"] }
swc_ecma_ast = "^15.0.0"
swc_ecma_codegen = "^16.0.0"
swc_ecma_parser = "^22.0.3"
swc_ecma_visit = "^14.0.0"
anyhow = { workspace = true }
glob = "0.3"
rayon = "1.8"
colored = "2.1"

[dev-dependencies]
tempfile = "3.9"
insta = "1.34"
pretty_assertions = "1.4"
criterion = { version = "0.5", features = ["html_reports"] }

[[bench]]
name = "formatting_bench"
harness = false

[[bench]]
name = "real_world_bench"
harness = false