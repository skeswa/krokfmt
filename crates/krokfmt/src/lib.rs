pub mod biome_formatter;
pub mod codegen;
pub mod comment_classifier;
pub mod comment_extractor;
pub mod comment_formatter;
pub mod comment_reinserter;
pub mod file_handler;
pub mod organizer;
pub mod parser;
pub mod selective_comment_handler;
pub mod semantic_hash;
pub mod transformer;

use anyhow::{Context, Result};
use std::path::Path;

/// Simple heuristic to detect JSX content in source code.
/// Looks for common JSX patterns like <Component> or JSX expressions.
fn contains_jsx(source: &str) -> bool {
    // Look for JSX element patterns: < followed by uppercase letter or lowercase HTML tag
    // This is a simple heuristic that covers most cases
    source.contains("</") || source.contains("/>") || 
    source.contains("React.") || source.contains("jsx") ||
    // Check for common JSX patterns
    source.chars().zip(source.chars().skip(1)).any(|(c1, c2)| {
        c1 == '<' && (c2.is_ascii_uppercase() || c2.is_ascii_lowercase())
    })
}

/// Format TypeScript/TSX code with krokfmt's opinionated rules.
///
/// This is the main entry point for programmatic use of krokfmt.
/// It applies the full formatting pipeline: parsing, organizing, and final formatting.
pub fn format_typescript(source: &str, filename: &str) -> Result<String> {
    // Auto-detect JSX content and use appropriate extension
    let has_jsx = contains_jsx(source);
    let effective_filename =
        if !filename.ends_with(".tsx") && !filename.ends_with(".jsx") && has_jsx {
            // If the filename doesn't already indicate JSX/TSX and we detected JSX, use .tsx
            "input.tsx".to_string()
        } else if filename.ends_with(".ts") && has_jsx {
            // If it's explicitly .ts but contains JSX, convert to .tsx
            filename.replace(".ts", ".tsx")
        } else {
            filename.to_string()
        };

    // Parse the TypeScript code
    let parser = parser::TypeScriptParser::new();
    let source_map = parser.source_map.clone();
    let comments = parser.comments.clone();
    let module = parser
        .parse(source, &effective_filename)
        .context("Failed to parse TypeScript code")?;

    // Organize the code structure with selective comment preservation
    let formatter = comment_formatter::CommentFormatter::new(source_map, comments);
    let organized_content = formatter
        .format(module, source)
        .context("Failed to organize code")?;

    // Apply final formatting with Biome
    let biome_formatter = biome_formatter::BiomeFormatter::new();
    let formatted_content = biome_formatter
        .format(&organized_content, Path::new(&effective_filename))
        .context("Failed to format with Biome")?;

    Ok(formatted_content)
}
