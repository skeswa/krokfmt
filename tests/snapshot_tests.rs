use insta::assert_snapshot;
use krokfmt::{codegen::CodeGenerator, formatter::KrokFormatter, parser::TypeScriptParser};
use std::fs;

fn format_code(input: &str) -> String {
    let parser = TypeScriptParser::new();
    let source_map = parser.source_map.clone();
    let comments = parser.comments.clone();
    let filename = if input.contains("<") && input.contains(">") {
        "test.tsx"
    } else {
        "test.ts"
    };
    let module = parser.parse(input, filename).unwrap();
    let formatted = KrokFormatter::new().format(module).unwrap();
    let generator = CodeGenerator::with_comments(source_map, comments);
    generator.generate(&formatted).unwrap()
}

fn test_fixture(fixture_path: &str) {
    let input_path = format!("tests/fixtures/{fixture_path}.input.ts");
    let input = fs::read_to_string(&input_path)
        .unwrap_or_else(|_| panic!("Failed to read fixture: {input_path}"));

    let output = format_code(&input);

    // Use the fixture path as the snapshot name
    assert_snapshot!(fixture_path, output);
}

// FR1: Import/Export Organization Tests

#[test]
fn test_fr1_1_default_imports() {
    test_fixture("fr1/1_1_default_imports");
}

#[test]
fn test_fr1_1_named_imports() {
    test_fixture("fr1/1_1_named_imports");
}

#[test]
fn test_fr1_1_namespace_imports() {
    test_fixture("fr1/1_1_namespace_imports");
}

#[test]
fn test_fr1_1_side_effect_imports() {
    test_fixture("fr1/1_1_side_effect_imports");
}

#[test]
fn test_fr1_1_type_imports() {
    test_fixture("fr1/1_1_type_imports");
}

#[test]
fn test_fr1_1_mixed_imports() {
    test_fixture("fr1/1_1_mixed_imports");
}

#[test]
fn test_fr1_1_import_aliases() {
    test_fixture("fr1/1_1_import_aliases");
}

#[test]
fn test_fr1_2_categorization() {
    test_fixture("fr1/1_2_categorization");
}

#[test]
fn test_fr1_3_sorting() {
    test_fixture("fr1/1_3_sorting");
}

#[test]
fn test_fr1_4_positioning() {
    test_fixture("fr1/1_4_positioning");
}

#[test]
fn test_fr1_5_group_separation() {
    test_fixture("fr1/1_5_group_separation");
}

#[test]
fn test_fr1_6_syntax_preservation() {
    test_fixture("fr1/1_6_syntax_preservation");
}

// FR2: Member Visibility Ordering Tests

#[test]
fn test_fr2_1_export_detection() {
    test_fixture("fr2/2_1_export_detection");
}

#[test]
fn test_fr2_2_export_prioritization() {
    test_fixture("fr2/2_2_export_prioritization");
}

#[test]
fn test_fr2_3_dependency_preservation() {
    test_fixture("fr2/2_3_dependency_preservation");
}

#[test]
fn test_fr2_3_circular_reference() {
    test_fixture("fr2/2_3_circular_reference");
}

#[test]
fn test_fr2_3_complex_chains() {
    test_fixture("fr2/2_3_complex_chains");
}

#[test]
fn test_fr2_3_interleaved_dependencies() {
    test_fixture("fr2/2_3_interleaved_dependencies");
}

#[test]
fn test_fr2_3_class_dependencies() {
    test_fixture("fr2/2_3_class_dependencies");
}

#[test]
fn test_fr2_3_namespace_dependencies() {
    test_fixture("fr2/2_3_namespace_dependencies");
}

#[test]
fn test_fr2_3_destructuring_dependencies() {
    test_fixture("fr2/2_3_destructuring_dependencies");
}

#[test]
fn test_fr2_3_computed_properties() {
    test_fixture("fr2/2_3_computed_properties");
}

#[test]
fn test_fr2_3_hoisting_challenges() {
    test_fixture("fr2/2_3_hoisting_challenges");
}

#[test]
fn test_fr2_3_forward_references() {
    test_fixture("fr2/2_3_forward_references");
}

#[test]
fn test_fr2_4_visibility_grouping() {
    test_fixture("fr2/2_4_visibility_grouping");
}

#[test]
fn test_fr2_4_visibility_with_dependencies() {
    test_fixture("fr2/2_4_visibility_with_dependencies");
}

#[test]
fn test_fr2_4_visibility_alphabetization() {
    test_fixture("fr2/2_4_visibility_alphabetization");
}

#[test]
fn test_fr2_1_export_detection_edge_cases() {
    test_fixture("fr2/2_1_export_detection_edge_cases");
}

#[test]
fn test_fr2_2_export_prioritization_complex() {
    test_fixture("fr2/2_2_export_prioritization_complex");
}

#[test]
fn test_fr2_4_visibility_mixed_declarations() {
    test_fixture("fr2/2_4_visibility_mixed_declarations");
}

#[test]
fn test_fr2_comprehensive_integration() {
    test_fixture("fr2/2_comprehensive_integration");
}

// FR3: Alphabetical Sorting Tests

#[test]
fn test_fr3_1_function_arguments() {
    test_fixture("fr3/3_1_function_arguments");
}

#[test]
fn test_fr3_2_object_properties() {
    test_fixture("fr3/3_2_object_properties");
}

#[test]
fn test_fr3_3_class_members() {
    test_fixture("fr3/3_3_class_members");
}

#[test]
fn test_fr3_3_visibility_groups() {
    test_fixture("fr3/3_3_visibility_groups");
}

#[test]
fn test_fr3_3_private_syntax() {
    test_fixture("fr3/3_3_private_syntax");
}

#[test]
fn test_fr3_3_mixed_visibility() {
    test_fixture("fr3/3_3_mixed_visibility");
}

#[test]
fn test_fr3_3_typescript_private() {
    test_fixture("fr3/3_3_typescript_private");
}

#[test]
fn test_fr3_4_type_members() {
    test_fixture("fr3/3_4_type_members");
}

#[test]
fn test_fr3_5_enum_members() {
    test_fixture("fr3/3_5_enum_members");
}

#[test]
fn test_fr3_6_jsx_properties() {
    test_fixture("fr3/3_6_jsx_properties");
}

// FR6: Comment Handling Tests

#[test]
fn test_fr6_1_line_comments() {
    test_fixture("fr6/6_1_line_comments");
}

#[test]
fn test_fr6_2_block_comments() {
    test_fixture("fr6/6_2_block_comments");
}

#[test]
fn test_fr6_3_jsdoc_comments() {
    test_fixture("fr6/6_3_jsdoc_comments");
}

#[test]
fn test_fr6_4_comment_positioning() {
    test_fixture("fr6/6_4_comment_positioning");
}

#[test]
fn test_fr6_5_comment_association() {
    test_fixture("fr6/6_5_comment_association");
}

#[test]
fn test_fr6_6_special_comments() {
    test_fixture("fr6/6_6_special_comments");
}

#[test]
fn test_fr6_7_standalone_comments() {
    test_fixture("fr6/6_7_standalone_comments");
}

// FR7: Visual Separation Tests

#[test]
fn test_fr7_1_module_separation() {
    test_fixture("fr7/7_1_module_separation");
}

#[test]
fn test_fr7_3_class_member_separation() {
    test_fixture("fr7/7_3_class_member_separation");
}
