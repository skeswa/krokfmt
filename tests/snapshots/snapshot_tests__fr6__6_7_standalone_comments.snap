---
source: tests/snapshot_tests.rs
expression: output
---
// FR6.7: Standalone Comment Preservation
// Test 1: Class with overall descriptive comment
// Test 5: Interface with documentation
interface APIClient {
    // Interface defining the contract for API clients
    // All implementations must support these methods

    get(url: string): Promise<any>;
    post(url: string, data: any): Promise<any>;
    delete(url: string): Promise<void>;
}
// Test 10: Method with contextual comments
class DataService {
    // Service configuration
    #apiUrl = 'https://api.example.com';
    #timeout = 5000;

    async fetchData(id: string) {
        // Validate input first
        if (!id) {
            throw new Error('ID required');
        }
        // Prepare the request
        // Note: We use a custom header for tracking
        const headers = {
            'Content-Type': 'application/json',
            'X-Request-ID': generateId()
        };
        try {
            // Make the actual request
            const response = await fetch(`${this.#apiUrl}/data/${id}`, {
                headers,
                timeout: this.#timeout
            });
            // Process the response
            // Important: Check status before parsing
            if (!response.ok) {
                throw new Error(`HTTP ${response.status}`);
            }
            return await response.json();
        } catch (error) {
            // Log and re-throw
            // This helps with debugging production issues
            console.error('Fetch failed:', error);
            throw error;
        }
    }
}
class DocumentProcessor {
    // This class handles document processing workflow
    // including validation, transformation, and storage

    static VERSION = '1.0.0';

    #privateConfig = {
        maxSize: 1024
    };

    processDocument(doc: any) {
        return doc;
    }
    validateDocument(doc: any) {
        return true;
    }
}
// Test 6: Complex class that triggered the issue
class MixedVisibilityClass {
    // Mix everything up to ensure proper sorting
    // Some public methods

    // Static public field
    static VERSION = '1.0.0';
    // Private static field

    static #config = {
        api: 'https://api.example.com'
    };
    // Public field

    name: string;
    // Another private field

    #id: number;

    zPublicMethod() {
        return 'z';
    }
    // Private method

    #validateName(name: string): boolean {
        return name.length > 0;
    }
}
// Test 4: Nested structures with class-level comments
class OuterClass {
    // This outer class contains nested functionality

    method() {
        // Method-level comment about implementation
        const helper = ()=>{
            // Arrow function with its own context
            return 42;
        };
        return helper();
    }

    #privateMethod() {
        // Private method implementation notes
        return 'private';
    }
}
// Test 3: Class with mixed comment types
class ServiceManager {
    // Overall service management logic
    // Handles lifecycle and dependency injection
    // Static configuration

    static #config = {
        timeout: 5000
    };

    name: string;
    // Instance fields

    #services = new Map();
    // Constructor initializes the manager

    constructor(name: string){
        this.name = name;
    }

    getService(id: string) {
        return this.#services.get(id);
    }
    // Public API
    registerService(id: string, service: any) {
        this.#services.set(id, service);
    }
}
// Test 9: Complex function with multiple section comments
function analyzeCodebase(files: string[]) {
    const results = {
        errors: [],
        stats: {},
        totalFiles: files.length,
        warnings: []
    };
    // Phase 1: Parse all files
    // This can take a while for large codebases
    const parsed = files.map((file)=>{
        try {
            return parseFile(file);
        } catch (e) {
            results.errors.push({
                error: e,
                file
            });
            return null;
        }
    }).filter(Boolean);
    // Phase 2: Build dependency graph
    // Note: We're using a simplified algorithm here
    // For production, consider using a more robust solution
    const depGraph = buildDependencyGraph(parsed);
    const cycles = detectCycles(depGraph);
    if (cycles.length > 0) {
        // Circular dependencies detected!
        // This is usually a code smell
        results.warnings.push(...cycles.map((c)=>({
                files: c,
                type: 'circular-dependency'
            })));
    }
    // Phase 3: Calculate metrics
    results.stats = {
        avgFileSize: calculateAvgSize(parsed),
        complexity: calculateComplexity(parsed),
        coverage: estimateCoverage(parsed)
    };
    // Cleanup temporary data
    // Important: This prevents memory leaks
    parsed.forEach((p)=>p.cleanup?.());
    return results;
}
function applyTransforms(data: number[], stats: any) {
    return data;
}
function buildDependencyGraph(parsed: any[]): any {
    return {};
}
function calculateAvgSize(parsed: any[]): number {
    return 0;
}
function calculateComplexity(parsed: any[]): number {
    return 0;
}
// Helper functions for tests
function calculateStats(data: number[]) {
    return {
        mean: 0,
        median: 0
    };
}
// Test 2: Function with implementation notes
function complexAlgorithm(data: number[]) {
    // This uses a two-pass approach:
    // 1. First pass calculates statistics
    // 2. Second pass applies transformations
    const stats = calculateStats(data);
    const sorted = data.sort();
    return applyTransforms(sorted, stats);
}
// Test 7: Function with section comments
function complexDataPipeline(rawData: any[]) {
    // Input validation
    const errors = [];
    for (const item of rawData){
        if (!isValid(item)) errors.push(item);
    }
    // Data transformation phase
    // We need to normalize before processing
    const normalized = rawData.map(normalize);
    const filtered = normalized.filter((x)=>x != null);
    // Performance optimization:
    // Batch process in chunks of 100
    const chunks = [];
    for(let i = 0; i < filtered.length; i += 100){
        chunks.push(filtered.slice(i, i + 100));
    }
    // Final aggregation
    return chunks.map(processChunk).flat();
}
// Test 8: Comments within function preserve their position
function dataProcessor(items: any[]) {
    // Initialize results container
    const results = [];
    // === VALIDATION PHASE ===
    // Check each item for required fields
    for (const item of items){
        if (!item.id || !item.name) {
            continue;
        }
        results.push(item);
    }
    // === TRANSFORMATION PHASE ===
    // Apply business logic transformations
    const transformed = results.map((item)=>{
        // Normalize the data structure
        return {
            ...item,
            normalized: true,
            timestamp: Date.now()
        };
    });
    // === FINAL CLEANUP ===
    return transformed.filter(Boolean);
}
function detectCycles(graph: any): any[] {
    return [];
}
function estimateCoverage(parsed: any[]): number {
    return 0;
}
// Helper functions for the examples
function generateId(): string {
    return Math.random().toString(36).substring(7);
}
function isValid(item: any): boolean {
    return item != null && item.id;
}
function normalize(item: any): any {
    return {
        ...item,
        normalized: true
    };
}
// Stub functions for the examples
function parseFile(file: string): any {
    return {
        ast: {},
        file
    };
}
function processChunk(chunk: any[]): any[] {
    return chunk.map((x)=>({
            ...x,
            processed: true
        }));
}
